internal/sourcemap/metadata.go

```diff
@@ -0,0 +1,409 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package sourcemap
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"sync"
+	"time"
+
+	"github.com/go-sourcemap/sourcemap"
+	"github.com/pkg/errors"
+
+	"github.com/elastic/apm-server/internal/elasticsearch"
+	"github.com/elastic/apm-server/internal/logs"
+	"github.com/elastic/elastic-agent-libs/logp"
+	"github.com/elastic/go-elasticsearch/v8/esapi"
+)
+
+type MetadataCachingFetcher struct {
+	esClient         *elasticsearch.Client
+	set              map[Identifier]string
+	alias            map[Identifier]*Identifier
+	mu               sync.RWMutex
+	backend          Fetcher
+	logger           *logp.Logger
+	index            string
+	init             chan struct{}
+	invalidationChan chan<- Identifier
+}
+
+func NewMetadataCachingFetcher(
+	c *elasticsearch.Client,
+	backend Fetcher,
+	index string,
+	invalidationChan chan<- Identifier,
+) *MetadataCachingFetcher {
+	return &MetadataCachingFetcher{
+		esClient:         c,
+		index:            index,
+		set:              make(map[Identifier]string),
+		alias:            make(map[Identifier]*Identifier),
+		backend:          backend,
+		logger:           logp.NewLogger(logs.Sourcemap),
+		init:             make(chan struct{}),
+		invalidationChan: invalidationChan,
+	}
+}
+
+func (s *MetadataCachingFetcher) Fetch(ctx context.Context, name, version, path string) (*sourcemap.Consumer, error) {
```

Comment:
The `MetadataCachingFetcher` should always only run as a background job, fetching metadata and populating a cache. Whereas the `Cache` and `Elasticsearch` Fetcher are responsible for synchr. fetching soucemap bodies from either the cache or ES. The `Fetch` method is called sync when processing incoming requests. Why exposing this `Fetch` from the `MetadataCachingFetcher`? This causes quite a strong coupling between the fetchers, especially the `MetadataCachingFetcher` doesn't need to know anything about the sourcemap body fetcher. It only needs to expose a way to access the cache with the metadata. 
I guess I had rather expected it the other way around, that the `MetadataCachingFetcher` would expose a public method `GetID` and the fetcher would be passed as a parameter to the `ElasticsearchFetcher`. 


Nit: How about also renaming the `ElasticsearchFetcher` to something indicating what it fetches from ES, as the `MetadataCachingFetcher` also fetches from ES.

Reply:
I've moved the logic for syncing and sending updates to a separate file and decoupled the fetchers.

Only the sync worker and the Elasticsearch fetchers are coupled/knows about ES. The fetchers have been renamed for clarity and they are decoupled from each other.
I'm not sure I fully understand where you are heading, do you want to merge Metadata and Body Caching fetchers ?